### DISCUSSION TITLE
How to create a package with dependencies that require reboots

### DISCUSSION THREAD

**Discussion by Grobi2010** [chocolatey/choco#2082] [General] [answered]
Hi @ferventcoder,
first of all, thanks for chocolatey, that saved me heaps and heaps of time..
But here is my current problem, eating up all my saved hours ;)

I'm completely lost, creating a package that has a dependency to dotnet4.x, kb2919355 and dotnet3.5. two of these dependencys need reboot after installation.
I tryed a lot to find out, how to handle the dependencys, but all my different solutions stuck on required reboots.
So I have no idea, how to maintain one of my packages (in this case garmin-express).

[https://chocolatey.org/packages/garmin-express](url) 

Any suggestion?

**Comment by ferventcoder** [comment, marked as answer]
### Chocolatey and Reboots

Chocolatey itself as a package manager will never perform a reboot - it's simply not in its purview to do that. However, it will tell you that you need to reboot, and with Chocolatey v0.10.12+ you can even set a feature flag for Chocolatey to stop execution if a needed reboot is detected (more on that in next section).

There are other tools that do infrastructure / configuration management where it would be in the purview of those tools to perform reboots. In those cases, those tools could perform reboots when needed. 

#### Exit codes / stop on need for reboot
One thing you can do that will help with locally running Chocolatey and integrations is have Chocolatey stop with certain exit codes that inform of a need for a reboot. They would even inform on what the next steps are. Those codes:

* 350/1604 - reboot, then run the same install/upgrade command - it should be idempotent
* 3010 - reboot and move on

The feature flag you are looking for is `exitOnRebootDetected`. From https://chocolatey.org/docs/chocolatey-configuration#exit-codes: 

> `[ ] exitOnRebootDetected` - Exit On Reboot Detected - Stop running install, upgrade, or uninstall when a reboot request is detected. Requires 'usePackageExitCodes' feature to be turned on. Will exit with either 350 or 1604. When it exits with 350, it means pending reboot discovered prior to running operation. When it exits with 1604, it means some work completed prior to reboot request being detected. Available in 0.10.12+.

`choco feature enable --name="'exitOnRebootDetected'"` (for a script) / `choco feature enable -n exitOnRebootDetected` for just running from powershell console.

To get the 3010 exit code, you may need to ensure `usePackageExitCodes` is turned on as well.

### Community Package Maintainer & Infrastructure
If you are maintaining packages and you add dependencies that need reboots, that's completely fine. As part of the instructions you receive by email, it lays out how to contain maintainers to request an exemption from the verifier that currently does not handle reboots.

So take the dependencies, then remember to reach back out to the moderators through the package review form to ask for an exemption due to x dependency requiring a reboot.

As of 02 SEP 2020, we are taking to pre-install some of the KB dependencies that would normally require reboots, so hopefully the need for handling reboots on the community won't be necessary as much.
